dot/telemetry/mailer.go

```diff
@@ -0,0 +1,161 @@
+	sync.Mutex
+}
+
+// Handler struct for holding telemetry related things
+type mailer struct {
+	messageQueue chan Message
+	connections  []*telemetryConnection
+	logger       log.LeveledLogger
+}
+
+func newMailer(logger log.LeveledLogger) *mailer {
+	return &mailer{
+		messageQueue: messageQueue,
+		logger:       logger,
+	}
+}
+
+// BootstrapMailer setup the mailer, the connections and start the async message shipment
+func BootstrapMailer(ctx context.Context, conns []*genesis.TelemetryEndpoint, logger log.LeveledLogger) error {
+	const retryDelay = time.Second * 15
+
+	mailer := newMailer(logger)
+
+	for _, v := range conns {
+		const maxRetries = 5
+		for connAttempts := 0; connAttempts < maxRetries; connAttempts++ {
+			conn, _, err := websocket.DefaultDialer.Dial(v.Endpoint, nil)
+			if err != nil {
+				mailer.logger.Debugf("cannot dial telemetry endpoint %s (try %d of %d): %s",
+					v.Endpoint, connAttempts+1, maxRetries, err)
+
+				timer := time.NewTimer(retryDelay)
+
+				select {
+				case <-timer.C:
+					continue
+				case <-ctx.Done():
+					mailer.logger.Debugf("bootstrap telemetry issue: %w", ctx.Err())
+					if !timer.Stop() {
+						<-timer.C
+					}
+
+					return ctx.Err()
+				}
+			}
+
+			mailer.connections = append(mailer.connections, &telemetryConnection{
+				wsconn:    conn,
+				verbosity: v.Verbosity,
+			})
+			break
+		}
+	}
+
+	if len(mailer.connections) == 0 {
+		return ErrInsufficientConnections
+	}
+
+	go mailer.asyncShipment(ctx)
+	return nil
+}
+
+// SendMessage sends Message to connected telemetry listeners through messageReceiver
+func SendMessage(msg Message) error {
+	const messageTimeout = time.Second
+
+	timer := time.NewTimer(messageTimeout)
+
+	select {
+	case messageQueue <- msg:
+		if !timer.Stop() {
+			<-timer.C
+		}
+
+	case <-timer.C:
+		return ErrTimoutMessageSending
+	}
+	return nil
+}
+
+func (m *mailer) asyncShipment(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case msg, ok := <-m.messageQueue:
+			if !ok {
+				return
+			}
+
+			go m.shipTelemetryMessage(msg)
+		}
+	}
+}
+
+func (m *mailer) shipTelemetryMessage(msg Message) {
+	msgBytes, err := msgToJSON(msg)
+	if err != nil {
+		m.logger.Debugf("issue encoding telemetry message: %s", err)
+		return
+	}
+
+	for _, conn := range m.connections {
+		conn.Lock()
+		defer conn.Unlock()
+
+		err = conn.wsconn.WriteMessage(websocket.TextMessage, msgBytes)
+		if err != nil {
+			m.logger.Debugf("issue while sending telemetry message: %s", err)
+		}
+	}
+}
+
+func msgToJSON(message Message) ([]byte, error) {
+	messageBytes, err := json.Marshal(message)
+	if err != nil {
+		return nil, err
+	}
+
+	messageMap := make(map[string]interface{})
+	err = json.Unmarshal(messageBytes, &messageMap)
+	if err != nil {
+		return nil, err
+	}
+
+	messageMap["ts"] = time.Now()
+	messageMap["msg"] = message.messageType()
+
+	fullRes, err := json.Marshal(messageMap)
+	if err != nil {
+		return nil, err
+	}
+	return fullRes, nil
```

Comment:
Try with


```go
type jsonMessage struct {
	Timestamp time.Time `json:"ts"`
	MessageType string `json:"msg"`
	*Message
}
```

That should put everything on the same level. Using `Message` instead of `*Message` doesn't properly inherit sorry I forgot